Test sync implementations

RoosterDragon opened the pull request:
Added tests to verify classes use both the ISync interface and the Sync attribute and don't forget one or the other when implementing syncable traits.

Fixed some traits that failed the tests.

RoosterDragon committed:
Add missing sync interfaces.

Mailaender commented:
I guess you can drop https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Lint/CheckSyncAnnotations.cs then.

Mailaender commented on OpenRA.Test/OpenRA.Game/SyncImplTest.cs:
```diff
@@ -0,0 +1,67 @@
+			foreach (var type in AllModTypes)
+				if (!TypeImplementsSync(type) && AnyTypeMemberIsSynced(type))
+					Assert.Fail("{0} has members with the Sync attribute but does not implement ISync".F(type.FullName));
+			Assert.Pass();
```
https://ci.appveyor.com/project/OpenRA/openra/build/1.0.5255/tests failed with an exception here.

RoosterDragon replied:
```
at NUnit.Framework.Assert.Pass(String message, Object[] args)
at NUnit.Framework.Assert.Pass()
at OpenRA.Test.SyncImplTest.TypesImplementingSyncInterfaceHaveSyncableMembers() in c:\projects\openra\OpenRA.Test\OpenRA.Game\SyncImplTest.cs:line 64
```

What does this mean? Does it think Assert.Pass() indicates a failure or something?

Mailaender commented on OpenRA.Test/OpenRA.Test.csproj:
```diff
@@ -43,12 +43,14 @@
       <Private>False</Private>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.Drawing" />
```
What do you need this for?

RoosterDragon replied:
For some reason it needs the reference in order to build. Otherwise I get these errors:
```
Error	CS0012	The type 'RectangleF' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.	OpenRA.Test	\OpenRA\OpenRA.Test\OpenRA.Game\SyncImplTest.cs	33
Error	CS0012	The type 'Rectangle' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.	OpenRA.Test	\OpenRA\OpenRA.Test\OpenRA.Game\SyncImplTest.cs	33
```

Phrohdoh replied:
Why are we using `RectangleF` is a sync test implementation?

RoosterDragon replied:
```return type.GetInterfaces().Contains(typeof(ISync))``` produces these errors (until you include the dll) - don't ask me!

A loop seems to avoid the problem - I'll use that.

RoosterDragon commented:
Didn't know there was a lint pass.

Should I implement it as a lint pass? That would allow 3rd party mods to check themselves too right?

Mailaender commented:
We just didn't have NUnit support back then. 3rd party mods should also have access to unit tests.

RoosterDragon commented:
Will the search for ```OpenRA.Mods.*.dll``` pick them up correctly?

RoosterDragon commented:
Updated:
- Also check and fix traits in OpenRA.Game.exe
- Remove the lint pass.
- Removed need to reference System.Drawing.

Mailaender commented on OpenRA.Test/OpenRA.Game/SyncImplTest.cs:
```diff
@@ -0,0 +1,71 @@
+			foreach (var type in AllTypes)
+				if (!TypeImplementsSync(type) && AnyTypeMemberIsSynced(type))
+					Assert.Fail("{0} has members with the Sync attribute but does not implement ISync".F(type.FullName));
+			Assert.Pass();
```
http://www.nunit.org/index.php?p=utilityAsserts&r=2.5

> The Assert.Pass method allows you to immediately end the test, recording it as successful. Since it causes an exception to be thrown, it is more efficient to simply allow the test to return.

RoosterDragon replied:
Ok, I'll just let it return then.

chrisforbes commented:
I like this *much* better as a lint pass, so you can apply it to a
particular set of DLLs.

On Thu, Jul 30, 2015 at 7:05 AM, RoosterDragon <notifications@github.com>
wrote:

> Didn't know there was a lint pass.
>
> Should I implement it as a lint pass? That would allow 3rd party mods to
> check themselves too right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/8839#issuecomment-126060847>.
>

Mailaender commented:
Why not just parametrize the DLL name?

RoosterDragon commented:
Perhaps I can do both (and have the test just run the lint pass or something like that).

Mailaender commented:
I am against a lint pass as it will run the test for each map (and this repository already has a lot of them) which is very wasteful and would wreck my current effort in reducing the Travis CI build times.

Mailaender commented:
These ISync annotation checks have nothing to do with MiniYaml. I added a workaround in https://github.com/OpenRA/OpenRA/pull/8936 to avoid doing duplicate checks for every map, but it just shows the insanity of using the inappropriate linter system.

RoosterDragon commented:
OK I've rebased and fixed the issues.

I agree with @Mailaender regarding the lint stuff - so I've stuck with using a test and removing the lint pass for now.

Mailaender commented:
Can you resubmit again? I had to manually increment the @AppVeyor build number again.

Mailaender commented:
:+1:

RoosterDragon commented:
I got an exception: "Actor type 'camera.paradrop' does not define a quantized body orientation." whilst running a game with this PR.

I suspect this is because the ```QuantizedFacings``` property in ```BodyOrientation``` is lazy and nothing was reading it before for this actor. However trying to sync it means it is now forcibly read and thus throwing the error.

Should we avoid syncing this? Or only sync it if something already generated it?

RoosterDragon commented:
I have decided to keep the current behaviour for ```BodyOrientation`` and avoid attempting to sync it since it may be problematic.

This resolves the crash.

penev92 commented:
Needs a rebase.

RoosterDragon commented:
Rebased, and also leave in syncing of ```BodyOrientation``` since #9193 fixed the underlying problem.

penev92 commented:
Sorry, needs a rebase again.

penev92 commented:
Sorry, needs a rebase again.

RoosterDragon commented:
Rebased.

pchote commented on OpenRA.Game/Actor.cs:
```diff
@@ -22,7 +22,7 @@
 namespace OpenRA
 {
-	public sealed class Actor : IScriptBindable, IScriptNotifyBind, ILuaTableBinding, ILuaEqualityBinding, ILuaToStringBinding, IEquatable<Actor>, IDisposable
+	public sealed class Actor : IScriptBindable, IScriptNotifyBind, ILuaTableBinding, ILuaEqualityBinding, ILuaToStringBinding, ISync, IEquatable<Actor>, IDisposable
```
I'm pretty sure we don't sync-test Actors (only traits), so this isn't very useful.

pchote replied:
We have a special-case path for syncing actors, but that only includes the actor ID.

RoosterDragon replied:
There is a [Sync] attribute on the player through, which seems like a reasonable thing to check. I'm kinda leaving it around as a mini-reminder to fix it eventually so we sync the actor correctly but until then I need ISync here to make the lint pass for now.

I could remove it if preferred.

pchote replied:
I'd prefer it if you could remove the [Sync] attribute from the owner, since it currently has no purpose except to be be misleading.

RoosterDragon replied:
Ok - done.

pchote commented on OpenRA.Test/OpenRA.Game/SyncImplTest.cs:
```diff
@@ -0,0 +1,69 @@
+	class SyncImplTest
+	{
+		static readonly IEnumerable<Type> AllTypes =
+			Directory.EnumerateFiles("mods", "OpenRA.Mods.*.dll", SearchOption.AllDirectories)
```
This isn't very nice: mods are free to load dlls from wherever they want (this becomes even more important after #8992).

RoosterDragon replied:
Where can I find the existing mechanism that does this loading? I'll just use that.

pchote replied:
The linter does it per-mod, using the mod.yaml definitions, but I don't think you can use that directly here.  Perhaps enumerate over all of the installed mods and build a list of all the possible mod dlls?

pchote replied:
This kind of trouble was why we decided to use lint passes for these things.

RoosterDragon committed:
Improved sync lint pass.

Now also checks for classes with Sync members lacking the ISync interface.

RoosterDragon commented:
OK we're back to a lint pass. It seems to be doing the right thing but I'm not super-sure it that means I have it implemented right.

Probably needs a review from fresh.

RoosterDragon committed:
Remove sync on actor - as it doesn't currently work.

pchote commented:
Thanks. Looks good to me now :+1:.  Confirmed that things break if I remove ISync or add it where it isn't needed.

penev92 commented:
:+1:
